### DISCUSSION TITLE
"dot" operator for complex types

### DISCUSSION THREAD

**Discussion by KrakenEngineer** [ncalc/ncalc#218] [General]
Hello
I use NCalc as part of my in-game programming language and i am trying to add binary "dot" operator to take an object or it's type and get fields/properties or invoke their methods where member names might be customized
Unfortunately, it seems not as easy as adding functions (or i didn't read the documentation well)
Is there a way to add "dot" operator? If it is, what is the simplest and safest one?

**Comment by gumbarros** [comment]
Hello @KrakenEngineer 

Like 1 . 1?

This would cause an error in our default parser, because dot is a decimal separator, but, we got you covered. You can write your own parser and use NCalc :).

You need to create your own [ILogicalExpressionFactory](https://ncalc.github.io/ncalc/api/NCalc.Factories.ILogicalExpressionFactory.html).

Check our default implementation:
https://github.com/ncalc/ncalc/blob/master/src/NCalc/Factories/LogicalExpressionFactory.cs

Then, you will use DI to add your LogicalExpressionFactory:
https://ncalc.github.io/ncalc/articles/dependency_injection.html

Edit:
You will also to create a custom [IEvaluationVisitor](https://ncalc.github.io/ncalc/api/NCalc.Visitors.IEvaluationVisitor.html). Use `BinaryExpressionType.Unknown` to handle the dot.

**Comment by KrakenEngineer** [comment]
Thanks!
Also i'd need to convert values from one type to another like "(bool)1" or "(int)1.2", is it possible to implement?

**Comment by gumbarros** [comment]
> Thanks! Also i'd need to convert values from one type to another like "(bool)1" or "(int)1.2", is it possible to implement?

Yes, you handle this at `IEvaluationVisitor`

**Comment by gmkado** [comment]
I'm not OP, but I wanted to do something similar.  Was able to achieve it by creating a custom LogicalExpressionParser and then replace [this line](https://github.com/ncalc/ncalc/blob/b9203fea4505ef60e7bdaae95a3f661ffa7de0f5/src/NCalc.Core/Parser/LogicalExpressionParser.cs#L128) with the following:

```cs
var identifier = Terms.Identifier(extraPart: c => c == '.');
```

Then create a custom LogicalExpressionFactory to use the custom LogicalExpressionParser.  Then I have an expression context that parses the '.' using the EvaluateParameterHandler.

I will say it's an unfortunate amount of copy-pasting to override this behavior.  @gumbarros would you be open to a PR that exposes the `extraPart` parameter of the identifier parser?  Or are you worried about exposing that (as I could see using the wrong characters might screw up parsing for other operations).
